File: src/main/scala/eu/unitn/disi/noise/generators/AbbreviationGen.scala
package eu.unitn.disi.noise.generators

import org.apache.spark.sql.expressions.UserDefinedFunction
import org.apache.spark.sql.functions.udf

class AbbreviationGen extends INoiseGen {

  override def description(): String = {
    s"${name()} keeps only the first char or digit of the input value"
  }

  override def name(): String = "ABBREVIATION"

  def generate(elem: String, rate: Double): Option[String] = {
    if (!toGenerate(rate)) {
      return Option(elem)
    }
    val str = Option(elem).getOrElse(return None)
    Some(str.substring(0, 1))
  }

  override def stringUDF(rate: Double): UserDefinedFunction = {
    udf[Option[String], String] { elem: String =>
      generate(elem, rate)
    }
  }

  def generate(elem: Int, rate: Double): Option[Int] = {
    if (!toGenerate(rate)) {
      return Option(elem)
    }
    val num = Option(elem).getOrElse(return None)
    if (num < 0) {
      Some(num.toString.substring(0, 2).toInt)
    } else {
      Some(num.toString.substring(0, 1).toInt)
    }
  }

  override def intUDF(rate: Double): UserDefinedFunction = {
    udf[Option[Int], Int] { elem: Int =>
      generate(elem, rate)
    }
  }

  def generate(elem: Long, rate: Double): Option[Long] = {
    if (!toGenerate(rate)) {
      return Option(elem)
    }
    val num = Option(elem).getOrElse(return None)
    if (num < 0) {
      Some(num.toString.substring(0, 2).toLong)
    } else {
      Some(num.toString.substring(0, 1).toLong)
    }
  }

  override def longUDF(rate: Double): UserDefinedFunction = {
    udf[Option[Long], Long] { elem: Long =>
      generate(elem, rate)
    }
  }

  def generate(elem: Double, rate: Double): Option[Double] = {
    if (!toGenerate(rate)) {
      return Option(elem)
    }
    val num = Option(elem).getOrElse(return None)
    if (num < 0) {
      Some(num.toString.substring(0, 2).toDouble)
    } else {
      Some(num.toString.substring(0, 1).toDouble)
    }
  }

  override def doubleUDF(rate: Double): UserDefinedFunction = {
    udf[Option[Double], Double] { elem: Double =>
      generate(elem, rate)
    }
  }
}

File: ng/generators/AbbreviationGen.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
import math
from .AbstractNoiseGen import AbstractNoiseGen
from pyspark.sql.types import *
from pyspark.sql import functions as F


class AbbreviationGen(AbstractNoiseGen):
    """
    This class is introduces abbreviations into the data.
    """

    def __init__(self, df, columns):
        super().__init__(df, columns)

    @staticmethod
    def description(**kwargs):
        return '{} keeps only the first char or digit of the input value' \
            .format(AbbreviationGen.name())

    @staticmethod
    def name(**kwargs):
        return 'ABBREVIATION'

    @staticmethod
    def abbreviation_str_generation(elem, distribution):
        if not distribution.generate(elem):
            if isinstance(elem, str):
                return elem
            return str(elem)
        if elem is None or not elem:
            return elem
        if isinstance(elem, str):
            return elem[0]
        return str(elem)[:1]

    @staticmethod
    def abbreviation_num_generation(elem, distribution):
        if not distribution.generate(elem):
            return elem
        if elem is None or elem == 0:
            return elem
        if elem < 0:
            return -(-elem // 10 ** int(math.log(-elem, 10)))
        return elem // 10 ** int(math.log(elem, 10))

    def string_udf(self, distribution):
        return F.udf(
            lambda elem:
            AbbreviationGen.abbreviation_str_generation(elem, distribution),
            StringType())

    def int_udf(self, distribution):
        return F.udf(
            lambda elem:
            AbbreviationGen.abbreviation_num_generation(elem, distribution),
            IntegerType())

    def double_udf(self, distribution):
        return F.udf(
            lambda elem:
            AbbreviationGen.abbreviation_num_generation(elem, distribution),
            DoubleType())

    def bigint_udf(self, distribution):
        return F.udf(
            lambda elem:
            AbbreviationGen.abbreviation_num_generation(elem, distribution),
            LongType())

    def tinyint_udf(self, distribution):
        return F.udf(
            lambda elem:
            AbbreviationGen.abbreviation_num_generation(elem, distribution),
            ByteType())

    def decimal_udf(self, distribution):
        return F.udf(
            lambda elem:
            AbbreviationGen.abbreviation_num_generation(elem, distribution),
            DecimalType())

    def smallint_udf(self, distribution):
        return F.udf(
            lambda elem:
            AbbreviationGen.abbreviation_num_generation(elem, distribution),
            ShortType())

    def date_udf(self, distribution):
        return F.udf(
            lambda elem:
            AbbreviationGen.abbreviation_str_generation(elem, distribution),
            StringType())

    def timestamp_udf(self, distribution):
        return F.udf(
            lambda elem:
            AbbreviationGen.abbreviation_str_generation(elem, distribution),
            StringType())

    def __str__(self):
        return '{} - {}'.format(AbbreviationGen.name(), self.columns)